### PR TITLE
Deprecate:Parent Span object from Datadog::Span

### PR DESCRIPTION
**Pull request by marcotc**
This PR removes the accessor `Datadog::Span#parent`, as well the writer `Datadog::Span#parent=` (alongside its alias `Datadog::Span#set_parent`). `Datadog::Span#parent_id` and `Datadog::Span#parent_id=`, which store the numerical parent span id are still available.

Modifying the parent after span creation is a perilous operation: the span carries multiple data points that are strongly tied to the parent, either directly or indirectly (trace_id, execution time, sampling rate, and service name in some case).
Setting an arbitrary parent at span creation time is always the advised route to avoid any possible internal span data inconsistencies.

The recommended way to do this is by passing `child_of` to `Datadog::Tracer.trace(name, child_of: parent_span)`.

If we still wish to manipulate the parent span object after span creation, `Datadog::SpanOperation#parent` and `Datadog::SpanOperation#parent=` are the new alternatives and should take care of all existing use cases we are aware of today.
Given that `Datadog::SpanOperation` represents a "live" span, modifying its parent has more well behaved semantics than modifying the parent object to an already finished `Datadog::Span`.

To summarize, the ability to get and set the `parent_id` remains unchanged (available in both `Span` and `SpanOperation`). The ability to set a `Datadog::Span` object as the parent span of an existing `Datadog::Span` is now only availalbe in a "live" span instance (`SpanOperation`). `Datadog::Tracer.trace(name, child_of: parent_span)` should always be used when possible, when there's a need for modifying the parent span.

### CODE REVIEW

**Comment by ivoanjo** [comment]
> This PR the accessor Datadog::Span#parent, as well the writer [...]

[I accidentally this PR lolol](https://knowyourmeme.com/memes/i-accidentally)

---

I'm curious about the decision to remove `#parent`. It does seem to make a lot of things more awkward.

```diff
-       expect(run_span.parent).to eq(rack_span)
+       expect(run_span.parent_id).to eq(rack_span.span_id)
```

For instance, where previously one could ask "is this span's parent the rack span?", we now need to go through a weird roundabout "is this span's parent id the same id as the rack span?".

Maybe these "queries" are not relevant/interesting enough to enable in our APIs...?

> To summarize, the ability to get and set the parent_id remains unchanged [...]

I'm also curious about the why on this decision.

I understand that it may be useful to be able to construct `Spans` with some set of data, but I don't quite understand the use case of after-creation mutation for ids -- wouldn't the same thinking that led us to remove `#parent=` apply to these as well?

**Comment by delner** [comment]
> For instance, where previously one could ask "is this span's parent the rack span?", we now need to go through a weird roundabout "is this span's parent id the same id as the rack span?".

Because `Span` effectively no longer holds a direct reference to its parent.

`SpanOperation`s are not `Span`s. `SpanOperation` produces a `Span`.

`SpanOperation` is intended to be the ephemeral, mutable, "unfinished" act of creating a span, which exposes references to other `SpanOperation`s to facilitate trace building. It's really only exposed as a facet of the `Tracer` API, but it shouldn't live outside the act of instrumentation.

`Span` is the completed measurement. It is what is sent to the `Writer` and is intended to be a serializable. Having object references on `Span` is neither necessary or desirable because it makes serialization more complicated.

These tests assert on `spans` which reads the completed `Span`s out of the test buffer, as they would have been written to the `Writer`. Hence, they can no longer assert the reference, they must assert the ID.

> but I don't quite understand the use case of after-creation mutation for ids -- wouldn't the same thinking that led us to remove #parent= apply to these as well?

`parent_id` shouldn't really ever be mutated after the span is created, even by the `Tracer`. Ideally, `Span` would be immutable entirely, only reason it isn't is because 1) `SpanOperation` creates & wraps a `Span` object and actively mutates it as a matter of convenience and 2) processing pipeline is a feature we offer.

That aside, I think there is actually a strong argument for not allowing any mutation of `parent_id` after `Span.new`. There's only one place I see in code that does this, which is the Ethon integration that does this to reflect a pretty trace. It may do that because of how its "multi" requests work. It might be worth changing that behavior so we can privatize `parent=`.

**Comment by marcotc** [comment]
> Ideally, `Span` would be immutable entirely, only reason it isn't is because 1) `SpanOperation` creates & wraps a `Span` object and actively mutates it as a matter of convenience and 2) processing pipeline is a feature we offer that allows mutation of `Span`.

1) Could be made private, and we, inside `ddtrace`, call that private `Span#set_parent` method. We'd hide it from the public API. This operation is not truly recommended (changing the parent after creation), so shouldn't happen frequently.

2) This is the one use case where I think there's a legitimate argument to keeping `Span#set_parent` public.

> That aside, I think there is actually a strong argument for not allowing any mutation of `parent_id` after `Span` or `SpanOperation` is created. There's only one place I see in code that does this, which is the Ethon integration that does this to reflect a pretty trace. It may do that because of how its "multi" requests work. It might be worth changing that behavior so we can privatize `parent=`.

This was removed in https://github.com/DataDog/dd-trace-rb/pull/1755.

File: lib/ddtrace/pipeline/span_filter.rb [hunk 1]
```diff
@@ -18,12 +18,12 @@ def call(trace)
...
-            deleted << span
+          if deleted.include?(span.parent_id)
+            deleted << span.span_id
             true
           else
             drop = drop_it?(span)
-            deleted << span if drop
+            deleted << span.span_id if drop
             drop
           end
```
Comment (ivoanjo) [inline lib/ddtrace/pipeline/span_filter.rb:28]: Tiny question: Would it make sense to make `deleted` a map (or a set) to enable faster `deleted.include?` lookups?

Also minor: I guess this could be simplified as

```ruby
trace.delete_if do |span|
  should_delete = deleted.include?(span.parent_id) || drop_it?(span)
  deleted << span.span_id if should_delete
  should_delete
end
```

File: lib/ddtrace/span_operation.rb [hunk 2]
```diff
@@ -189,6 +182,22 @@ def finished?
...
 
+    # Set this span's parent, inheriting any properties not explicitly set.
+    # If the parent is nil, set the span as the root span.
+    def parent=(parent)
+      @parent = parent
+
+      if parent.nil?
+        span.trace_id = span.span_id
+        span.parent_id = 0
+      else
```
Comment (ivoanjo) [inline lib/ddtrace/span_operation.rb:193]: Minor: I wonder if it'd make sense to make this more explicit. E.g. have a deliberate and explicit `make_root_span` method if one really wants to make a span into a root, and otherwise complain if someone tries to do it via setting `parent = nil`.

File: spec/ddtrace/contrib/concurrent_ruby/integration_test_spec.rb [hunk 3]
```diff
@@ -46,8 +46,8 @@
       deferred_execution
     end
 
-    it 'creates outer span with nil parent' do
-      expect(outer_span.parent).to be_nil
+    it 'creates outer span without a parent' do
+      expect(outer_span.parent_id).to eq(0)
     end
```
Comment (ivoanjo) [inline spec/ddtrace/contrib/concurrent_ruby/integration_test_spec.rb:51]: I wonder if it would make sense to explicitly name this concept; e.g. create `Span#root_span?`.

Then, we could make all these tests a lot more explicit: `expect(outer_span).to be_root_span`

---

We could even go as far as also having `Span#has_parent?(span) = parent_id == span.span_id` and then also do `expect(some_span).to have_parent?(another_span)`.

File: lib/ddtrace/span_operation.rb [hunk 4]
```diff
@@ -189,6 +182,22 @@ def finished?
       span.stopped?
     end
 
+    # Set this span's parent, inheriting any properties not explicitly set.
+    # If the parent is nil, set the span as the root span.
+    def parent=(parent)
```
Comment (ivoanjo) [inline lib/ddtrace/span_operation.rb:187]: Important: I just noticed that `parent=` changes state on both the `SpanOperation` and the `Span`; but `parent_id=` only changes the `Span` and will leave the `SpanOperation` in an inconsistent state, because `@parent` will be incorrect.

**Review by ivoanjo — COMMENTED** [review]
Left a few more comments. Thanks y'all for explaining things to the noob :)

> parent_id shouldn't really ever be mutated after the span is created, even by the Tracer. Ideally, Span would be immutable entirely, only reason it isn't is because 1) SpanOperation creates & wraps a Span object and actively mutates it as a matter of convenience and 2) processing pipeline is a feature we offer that allows mutation of Span.

I wonder if it would make sense to move all of the mutation to `SpanOperation`; since the `SpanOperation` just wraps the `Span`, it looks like we could do it without impacting any of the callers/users, right?

> 2) processing pipeline is a feature we offer that allows mutation of Span.` 

> This is the one use case where I think there's a legitimate argument to keeping Span#set_parent public.

Have we considered making it a persistant data structure, rather than just immutable?

E.g. instead of just disallowing mutation of `Span`s:

```
[2] pry(main)> array = [1, 2, 3].freeze
=> [1, 2, 3]
[3] pry(main)> array << 4
FrozenError: can't modify frozen Array: [1, 2, 3]
```

we could make would-be mutations just return a new span, with the changed attributes:

```
[6] pry(main)> require 'persistent_dmnd/everywhere' # persistent-dmnd gem
[7] pry(main)> array = a💎[1, 2, 3]
=> Persistent💎::Array[1, 2, 3]
[8] pry(main)> array << 4
=> Persistent💎::Array[1, 2, 3, 4] # => new object with added element
[9] pry(main)> array
=> Persistent💎::Array[1, 2, 3] # => original object is retained
```

File: lib/ddtrace/pipeline/span_filter.rb (hunk 1 above)
Comment (marcotc) [reply lib/ddtrace/pipeline/span_filter.rb:28]: 👍

File: lib/ddtrace/span_operation.rb (hunk 4 above)
Comment (marcotc) [reply lib/ddtrace/span_operation.rb:187]: Yeah, that one the reasons we are moving to get rid of `parent`.

There's also the fact that for distributed spans (the ones that have a parent outside the of the current application), we can only ever set the `parent_id` never the `parent` because a parent Span object doesn't exist locally.

This PR doesn't remove `parent` completely, but that's the final goal.

File: spec/ddtrace/contrib/concurrent_ruby/integration_test_spec.rb (hunk 3 above)
Comment (marcotc) [reply spec/ddtrace/contrib/concurrent_ruby/integration_test_spec.rb]: I added `:be_root_span`.

Regarding `have_parent?`, I'd rather not expand our public API (by adding `Span#has_parent?`). And because we can't make it private, and package visibility is not a thing in Ruby, I'm not sure we have it as suggested.

Would you think it makes sense do add a custom `has_parent?` RSpec patcher instead?

File: lib/ddtrace/span_operation.rb (hunk 2 above)
Comment (marcotc) [reply lib/ddtrace/span_operation.rb:193]: Yes, this method's API is super convoluted.
I think we actually want to remove it altogether, specially this special `nil` case.
I'll document this.

**Review by delner — APPROVED** [review]

File: spec/ddtrace/contrib/concurrent_ruby/integration_test_spec.rb (hunk 3 above)
Comment (ivoanjo) [reply spec/ddtrace/contrib/concurrent_ruby/integration_test_spec.rb]: > Would you think it makes sense do add a custom has_parent? RSpec patcher instead?

Yeah I guess it would be a nice to have.
